DOCSP-29510: explain how driver handles undefined vals (#674)
mongodb/docs-node: source/fundamentals/bson/undefined-values.txt

.. _node-undefined-values:

================
Undefined Values
================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to control how the driver serializes
undefined values. By default, the driver serializes ``undefined`` values
as ``null`` values during write operations.

.. _nodejs-specify-ignoreundefined:

> Specify the ``ignoreUndefined`` Setting

Reviewer: Suggestion: I think this section could potentially be renamed something like "Ignore Undefined Values" since that's the core of what they are learning how to do here.

     state: "New Hampshire",
     salesTax: null
   }
   
The driver does not serialize the ``salesTax``
field during the first insert operation because we specified the
``ignoreUndefined`` setting.

.. _nodejs-ignoreundefined-scope:

Set Scope for Serializing Undefined Values
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can specify the ``ignoreUndefined`` setting at the client,
database, collection, or operation level.

The ``ignoreUndefined`` setting automatically applies to the scope of the
object instance in which you specified it, and any other objects created
from that instance.

For example, if you include the option when instantiating a database
object, any collection instance you extract from that object inherits
the setting. Any operations you call on that collection instance also
inherit the setting.

.. code-block:: javascript

   const myDB = client.db("test", { ignoreUndefined: true });

   // The collection inherits the ignoreUndefined setting
   const myColl = myDB.collection("states");

   // Any write operation will not serialize undefined values
   await myColl.findOneAndUpdate(
     { state: "Georgia" },
     { $set: { gasTax: undefined } }
   );

> You can override the setting at any level of scope by including it when
> constructing the object instance or when executing an operation.

Reviewer: I found this a bit hard to parse out. It could possibly be simplified to something like this, since there is an example immediately following.

```suggestion
You can specify ``ingoreUndefined` again at any level to override the inherited settings.
```

For example, if you set ``ignoreUndefined`` to ``true`` on your
collection object, you can override the setting in individual write
operations that you execute on that collection.

> .. code-block:: javascript

Author: technically, this update operation would not cause any changes to the data. I can include that somewhere in the text, if you think it would be useful for users to know

.. _node-undefined-values:

================
Undefined Values
================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn to control how the driver serializes
``undefined`` values. By default, the driver serializes ``undefined`` values
as ``null`` values during write operations.

.. _nodejs-specify-ignoreundefined:

Ignore Undefined Values
-----------------------

To make the driver ignore fields with
``undefined`` values during serialization, set the
``ignoreUndefined`` setting to ``true``. When you specify this setting,
the driver *does not* serialize fields with ``undefined`` values.

> The following example inserts two documents. One insert operation has
> the ``ignoreUndefined`` setting set to ``true``. The driver does not
> serialize the ``salesTax`` field during the first insert operation
> because we specified the ``ignoreUndefined`` setting:

Reviewer: Suggestion: I think these sentences can be combined:

```suggestion
The following example inserts two documents. The first insert operation has
the ``ignoreUndefined`` setting set to ``true``, so it does not
serialize the ``salesTax`` field.
```
